src/persist-client/src/cfg.rs

```diff
@@ -278,17 +259,23 @@ impl PersistConfig {
 
         let mut cfg = Self::new(&DUMMY_BUILD_INFO, SYSTEM_TIME.clone());
         cfg.hostname = "tests".into();
-        cfg.dynamic
-            .set_feature_flag(PersistFeatureFlag::STREAMING_COMPACTION, true);
-        cfg.dynamic
-            .set_feature_flag(PersistFeatureFlag::STREAMING_SNAPSHOT_AND_FETCH, true);
+        cfg.set_config(&STREAMING_COMPACTION_ENABLED, true);
+        cfg.set_config(&STREAMING_SNAPSHOT_AND_FETCH_ENABLED, true);
         cfg
     }
 }
 
 #[allow(non_upper_case_globals)]
 pub(crate) const MiB: usize = 1024 * 1024;
 
+/// Returns the full set of all persist [Config]s.
+pub fn all_dyn_configs() -> ConfigSet {
```

Comment:
This is a third touchpoint when adding a flag in addition to the two I mention in the "theoretical minimum" section.

After letting this mull, I'm a pretty strong yes on replacing it with a static global registry at this point. Basically, the boilerplate has gotten light enough that any additional work now feels meaningful in comparison. However not doing that in this PR because (1) IIUC Ben is a skeptic of the global state and (2) I'm not sure I understand how sketch the crates available to do that are.

For (1), while I strongly agree that flag _values_ should not be global state, I have a harder time making an argument that global state containing the set of flag _definitions_ in a binary is something to worry about in practice. However, the one worry I do have here is that a config defined in code that exists only in clusterd (e.g. much of compute) wouldn't get picked up by envd and the explicit declarations currently here would prevent that subtlety.

For (2), we do have one of the possibly-sketchy crates [already used in mz_ore](https://github.com/MaterializeInc/materialize/blob/d11c694ef8d207c74ad8afa1aa0a026de02f0a20/src/ore/src/assert.rs#L60). I proved to myself in a local branch that we could use this one already in the dep graph for configs as long as it's possible to use a Mutex in the `ctor` section. There's also something like the `inventory` crate which is essentially designed exactly for this registration sort of use.

Reply:
> IIUC Ben is a skeptic of the global state

Yeah, still true! Basically I find the idea of code's behaviour changing depending on what's happens to be linked in to be sort of alarming and hard to reason about. Like, suppose I hit the `clusterd` bug you're talking about by taking a function with only one caller and moving it next to the code that calls it; normally a harmless refactoring, but now I've just accidentally changed materialize's SQL API. Spooky!

Conversely, I think the one extra line just bothers me less. (Which I appreciate is a matter of taste!)

Reply:
> However, the one worry I do have here is that a config defined in code that exists only in clusterd (e.g. much of compute) wouldn't get picked up by envd and the explicit declarations currently here would prevent that subtlety.

Hm, I was also in favor of the global flag registry until you mentioned this footgun. That seems somewhat hard to work around? 

One idea that comes to mind: when `envd` initializes its connection with `clusterd`, it sends over all the list of all known flags. If `clusterd` knows of any flags that `envd` doesn't send over, then `clusterd` panics with an informative message. That seems possibly workable!

Reply:
> don't think we need to block on the specifics here

Works for me!